gnosis/verification.k

```diff
@@ -99,31 +116,69 @@ module VERIFICATION
     rule 0 <=Int #bufElm(_, _)  => true
     rule #bufElm(_, _) <Int 256 => true
 
+    rule 0 <=Int X xorInt 115792089237316195423570985008687907853269984665640564039457584007913129639935 => true      requires 0<=Int X andBool X <Int pow256
+
+    rule X xorInt 115792089237316195423570985008687907853269984665640564039457584007913129639935 <Int pow256 => true  requires 0<=Int X andBool X <Int pow256 
+
+
   // ########################
   // Simplification
   // ########################
 
+    rule (X +Int 31) /Int 32 *Int 32 => #ceil32(X)
+    rule (31 +Int X) /Int 32 *Int 32 => #ceil32(X)
+
+    rule #asWord(#buf(32, DATA)) => DATA  requires #rangeUInt(256, DATA)
+
     rule #padToWidth(32, #asByteStack(V)) => #buf(32, V) // #asByteStackInWidth(V, 32)
-      requires 0 <=Int V andBool V <Int pow256
+      requires 0 <=Int V andBool V <Int pow256 andBool #getKLabelString(V) =/=String "#asWord"
 
+    rule #buf(32, #asWord(WS)) => WS  requires #sizeWordStack(WS) ==Int 32
     rule #buf(32, #asWord(#bufSeg(BUF, START, WIDTH))) => #bufSeg(BUF, START +Int WIDTH -Int 32, 32)  requires WIDTH >=Int 32
 
     rule #asWord(#bufSeg(BUF, START, WIDTH)) &Int 255 => #asWord(#bufSeg(BUF, START +Int WIDTH -Int 1, 1))  requires WIDTH >=Int 1
     rule 255 &Int #asWord(#bufSeg(BUF, START, WIDTH)) => #asWord(#bufSeg(BUF, START +Int WIDTH -Int 1, 1))  requires WIDTH >=Int 1
 
     rule #bufSeg(BUF, START, 1) => #bufElm(BUF, START) : .WordStack
+    rule #asWord(#bufElm(BUF, INDEX) : .WordStack) => #bufElm(BUF, INDEX)
+
+    rule #buf(32, DATA) => #padToWidth(32, #asByteStack(DATA))  [concrete]
 
     rule #asWord(WS0 ++ #buf(MASKWIDTH0, 0)) |Int #asWord(#buf(MASKWIDHT1, 0) ++ WS1)
       => #asWord(WS0 ++ WS1)
       requires #sizeWordStack(WS0) +Int MASKWIDTH0 ==Int 32
        andBool #sizeWordStack(WS0) ==Int MASKWIDHT1
        andBool MASKWIDTH0 ==Int #sizeWordStack(WS1)
 
-    // Arithmetic
+    rule #asWord(WS) &Int 115792089237316195423570985008687907853269984665640564039457584007913129639935 => #asWord(WS)
+
+    // 2^256 - 2^160 = 0xff..ff00..00 (96 1's followed by 160 0's)
+    rule 115792089237316195423570985007226406215939081747436879206741300988257197096960 &Int ADDR => 0
+      requires #rangeAddress(ADDR)
+
+    rule 1 |Int chop(A xorInt maxUInt256) => 1
+
+    rule chop((ADDR &Int maxUInt160) modInt pow160) => ADDR
+      requires #rangeAddress(ADDR)
+
+
+  // ########################
+  // Arithmetic
+  // ########################
+
     rule (A +Int I2) +Int (I3 -Int A) => I2 +Int I3
     rule (A +Int I2) -Int (A +Int I3) => I2 -Int I3
     rule (A +Int I2) -Int (I3 +Int A) => I2 -Int I3
 
+    rule (I1 +Int I2) -Int (I3 +Int I1) => I2 -Int I3
+
+    rule A -Int (A -Int I1) => I2
+
+    // Simplify Cmem(_,_) - Cmem(_, _)
+    rule A +Int (I1 -Int A) => I1
+    rule (I1 +Int (A -Int I0)) +Int (I2 -Int A) => I1 +Int (I2 -Int I0)
```

Comment:
don't we need to put more side condition so that these rules are applied in a controlled way?
e.g., isConcrete(I)  and/or  getKLabel(A) == "Cmem" ...?

Reply:
Not sure. Eliminating the common expression is always preferred, isn't it?

Reply:
maybe you're right in this case.  i'm talking about some general principle.